h2o-genmodel-extensions/xgboost/src/main/java/hex/genmodel/algos/xgboost/XGBoostNativeMojoModel.java

```diff
@@ -0,0 +1,171 @@
+            _nclasses, _priorClassDistrib, _defaultThreshold, _sparse);
+  }
+
+  public static double[][] bulkScore0(double[][] doubles, double[] offsets, double[][] preds,
+                                      Booster _booster, int _nums, int _cats,
+                                      int[] _catOffsets, boolean _useAllFactorLevels,
+                                      int nclasses, double[] _priorClassDistrib,
+                                      double _defaultThreshold, boolean _sparse) {
+    if (offsets != null) throw new UnsupportedOperationException("Unsupported: offset != null or only 0s");
+    float[][] floats;
+    int cats = _catOffsets == null ? 0 : _catOffsets[_cats];
+    // convert dense doubles to expanded floats
+    floats = new float[doubles.length][_nums + cats]; //TODO: use thread-local storage
+    for(int i = 0; i < doubles.length; i++) {
+      GenModel.setInput(doubles[i], floats[i], _nums, _cats, _catOffsets, null, null, _useAllFactorLevels, _sparse /*replace NA with 0*/);
+    }
+    float[][] out = null;
+    DMatrix dmat = null;
+    try {
+      Map<String, String> rabitEnv = new HashMap<>();
+      rabitEnv.put("DMLC_TASK_ID", "0");
+      Rabit.init(rabitEnv);
+      dmat = new DMatrix(floats,doubles.length,floats[0].length, _sparse ? 0 : Float.NaN);
+//      dmat.setWeight(new float[]{(float)weight});
+      out = _booster.predict(dmat);
+      Rabit.shutdown();
+    } catch (XGBoostError xgBoostError) {
+      throw new IllegalStateException("Failed XGBoost prediction.", xgBoostError);
+    } finally {
+      BoosterHelper.dispose(dmat);
+    }
+
+    for(int r = 0; r < out.length; r++) {
+      if (nclasses > 2) {
+        for (int i = 0; i < out[0].length; ++i)
+          preds[r][1 + i] = out[r][i];
+//      if (_balanceClasses)
+//        GenModel.correctProbabilities(preds, _priorClassDistrib, _modelClassDistrib);
+        preds[r][0] = GenModel.getPrediction(preds[r], _priorClassDistrib, doubles[r], _defaultThreshold);
+      } else if (nclasses == 2) {
+        preds[r][1] = 1 - out[r][0];
+        preds[r][2] = out[r][0];
+        preds[r][0] = GenModel.getPrediction(preds[r], _priorClassDistrib, doubles[r], _defaultThreshold);
+      } else {
+        preds[r][0] = out[r][0];
+      }
+    }
+    return preds;
+  }
+
+  public static double[] score0(double[] doubles, double offset, double[] preds,
+                                Booster _booster, int _nums, int _cats,
+                                int[] _catOffsets, boolean _useAllFactorLevels,
+                                int nclasses, double[] _priorClassDistrib,
+                                double _defaultThreshold, boolean _sparse) {
+    if (offset != 0) throw new UnsupportedOperationException("Unsupported: offset != 0");
+    float[] floats;
+    int cats = _catOffsets == null ? 0 : _catOffsets[_cats];
+    // convert dense doubles to expanded floats
+    floats = new float[_nums + cats]; //TODO: use thread-local storage
```

Comment:
Should we resolve the TODO before merging ?

Reply:
In this case, I think it is not necessary (couldn't be harmful), this was there before - just moving code around. Allocating this array doesn't really hurt the performance in XGBoost - there is a lot of other stuff going on.